### PR TITLE
add documentation comments to dockerfile

### PR DESCRIPTION
**Pull request by webern**
Some separation and comments in the Dockerfile to help those who may be
reading it for the first time.

**Issue number:**

N/A

**Description of changes:**

I'm trying to better understand the build process and found it difficult to get a foothold in the Dockerfile. I think this may help the next person.

Explaining what `RUN --mount=target=/host` means at the top might seem weird, but I spent a lot of time digging to try and figure that out. Eventually I had to test it in a minimal Dockerfile to actually get it (the documentation I found didn't fully illuminate that buidlkit feature for me).

Adding the cat separators before each `FROM` makes it easier to scroll back and reference the previous build stage.

**Testing done:**

N/A

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by srgothi92 — APPROVED** [review]
💯

**Review by etungsten — APPROVED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+# build context (which in practice is the root of the Bottlerocket repository)
+# as a read-only filesystem at /host.
 
 ARG SDK
 FROM ${SDK} as sdk
 
+################################################################################################
+# Section 1: The following build stages are used to build rpm.spec packages
+
+# ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^ ^..^
```
Comment (etungsten) [inline Dockerfile]: small nit: missing whiskers :cat:  =^..^=

Comment (webern) [reply Dockerfile]: Oh yeah, oops.

**Comment by webern** [comment]
> webern force-pushed the webern:dockerfile-doc branch from 52c5311 to cfc9bf6 20 seconds ago

I gave whiskers to the cats.

The lack of wrap-width consistency is driving me a little bit crazy. I don't suppose there's anything we can do about those long --mount lines, but everything else could be hard wrapped/re-wrapped at some width.  80 is a bit short for this file, but 100 would be a bit wide. Lines 18-22 are wrapped at about 90 which feels right.

File: Dockerfile [hunk 2]
```diff
@@ -88,9 +103,18 @@ RUN --mount=source=.cargo,target=/home/builder/.cargo \
...
 
+# =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^=
+# Copies RPM packages from the previous stage to their expected location so
+# that buildsys can find them and copy them out.
 FROM scratch AS package
 COPY --from=rpmbuild /home/builder/rpmbuild/RPMS/*/*.rpm /output/
 
+############################################################################################
+# Section 2: The following build stages are used to create a Bottlerocket image
+# once all of the rpm files have been created by repeatedly using of Section 1.
```
Comment (tjkirch) [inline Dockerfile]:
```suggestion
# once all of the rpm files have been created by repeatedly using Section 1.
```

**Review by tjkirch — APPROVED** [review]

**Review by zmrow — APPROVED** [review]
:vulcan_salute:

**Review by samuelkarp — APPROVED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -120,6 +144,8 @@ RUN --mount=target=/host \
     && createrepo_c /local/rpms \
     && echo ${NOCACHE}
 
+# =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^=
+# Builds a bottlerocket image.
```
Comment (bcressey) [inline Dockerfile]:
```suggestion
# Builds a Bottlerocket image.
```

**Review by bcressey — COMMENTED** [review]
> The lack of wrap-width consistency is driving me a little bit crazy. I don't suppose there's anything we can do about those long --mount lines, but everything else could be hard wrapped/re-wrapped at some width. 80 is a bit short for this file, but 100 would be a bit wide. Lines 18-22 are wrapped at about 90 which feels right.

It's fine with me if you want to fix that here. I would just ask that you not re-expand the command lines that are already split up - like `createrepo_c` and `dnf -y` - since I find the current format more readable. But the comments and the very long `find` commands are fair game.

**Comment by webern** [comment]
> webern force-pushed the webern:dockerfile-doc branch from cfc9bf6 to b8646f6 1 minute ago

I placed a ruler on the right-most cat whisker and wrapped things that overhung it. I rewrapped comments that were at a shorter width.

To me this is much prettier. Unfortunately the RUN --mount lines that are long cannot be wrapped and have their next line indented (no spaces allowed), but I think even this less-ideal wrap is better than hanging off the edge of the world.

File: Dockerfile [hunk 4]
```diff
@@ -101,7 +124,9 @@ USER root
 RUN --mount=target=/host \
     mkdir -p /local/rpms /local/migrations ./rpmbuild/RPMS \
     && ln -s /host/build/rpms/*.rpm ./rpmbuild/RPMS \
-    && find /host/build/rpms/ -maxdepth 1 -type f -name "bottlerocket-${ARCH}-migrations-*.rpm" -not -iname '*debuginfo*' -exec cp '{}' '/local/migrations/' ';' \
+    && find /host/build/rpms/ -maxdepth 1 -type f \
+        -name "bottlerocket-${ARCH}-migrations-*.rpm" \
+        -not -iname '*debuginfo*' -exec cp '{}' '/local/migrations/' ';' \
```
Comment (tjkirch) [inline Dockerfile]:
```suggestion
        -not -iname '*debuginfo*' \
        -exec cp '{}' '/local/migrations/' ';' \
```
nit: the exec is critically different than the find criteria.

File: Dockerfile [hunk 5]
```diff
@@ -78,19 +90,30 @@ RUN --mount=target=/host \
...
-# the variant-specific models.
+# We use the "nocache" writable space to generate code where necessary, like the variant-
+# specific models.
 USER builder
 RUN --mount=source=.cargo,target=/home/builder/.cargo \
     --mount=type=cache,target=/home/builder/.cache,from=cache,source=/cache \
-    --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/models/src/variant,from=variantcache,source=/variantcache \
-    --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/logdog/conf/current,from=variantcache,source=/variantcache \
+    --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/models/src/variant,\
+from=variantcache,source=/variantcache \
```
Comment (tjkirch) [inline Dockerfile]: This looks really odd to me.  If the long-ish line really bugs you, I guess I'd suggest making the `target` equivalent to the `from` on its own new line..?

**Review by tjkirch — APPROVED** [review]

**Comment by webern** [comment]
> webern force-pushed the webern:dockerfile-doc branch from b8646f6 to 983cb52 1 minute ago

I restored the --mount lines to their less-controversial long form, and took @tjkirch's other suggestion.

**Review by zmrow — APPROVED** [review]
:octopus:

**Review by tjkirch — APPROVED** [review]

**Pull request merged**